nifi-nar-bundles/nifi-standard-services/nifi-distributed-cache-services-bundle/nifi-distributed-cache-server/src/main/java/org/apache/nifi/distributed/cache/server/codec/MapCacheRequestDecoder.java

```diff
@@ -0,0 +1,153 @@
+import java.util.Optional;
+import java.util.OptionalInt;
+import java.util.OptionalLong;
+
+/**
+ * Cache Request Decoder processes bytes and decodes cache version and operation requests
+ */
+public class MapCacheRequestDecoder extends CacheRequestDecoder {
+
+    public MapCacheRequestDecoder(
+            final ComponentLog log,
+            final int maxLength,
+            final CacheOperation[] supportedOperations
+    ) {
+        super(log, maxLength, supportedOperations);
+    }
+
+    @Override
+    protected Object readRequest(final CacheOperation cacheOperation, final ByteBuf byteBuf) {
+        final Object request;
+
+        if (MapOperation.CONTAINS_KEY == cacheOperation) {
+            request = readKeyRequest(cacheOperation, byteBuf);
+        } else if (MapOperation.FETCH == cacheOperation) {
+            request = readKeyRequest(cacheOperation, byteBuf);
+        } else if (MapOperation.GET == cacheOperation) {
+            request = readKeyRequest(cacheOperation, byteBuf);
+        } else if (MapOperation.GET_AND_PUT_IF_ABSENT == cacheOperation) {
+            request = readKeyValueRequest(cacheOperation, byteBuf);
+        } else if (MapOperation.KEYSET == cacheOperation) {
+            request = new MapCacheRequest(cacheOperation);
+        } else if (MapOperation.REMOVE == cacheOperation) {
+            request = readKeyRequest(cacheOperation, byteBuf);
+        } else if (MapOperation.REMOVE_BY_PATTERN == cacheOperation) {
+            request = readPatternRequest(cacheOperation, byteBuf);
+        } else if (MapOperation.REMOVE_BY_PATTERN_AND_GET == cacheOperation) {
+            request = readPatternRequest(cacheOperation, byteBuf);
+        } else if (MapOperation.REMOVE_AND_GET == cacheOperation) {
+            request = readKeyRequest(cacheOperation, byteBuf);
+        } else if (MapOperation.REPLACE == cacheOperation) {
+            request = readKeyRevisionValueRequest(cacheOperation, byteBuf);
+        } else if (MapOperation.SUBMAP == cacheOperation) {
+            request = readSubMapRequest(cacheOperation, byteBuf);
+        } else if (MapOperation.PUT == cacheOperation) {
+            request = readKeyValueRequest(cacheOperation, byteBuf);
+        } else if (MapOperation.PUT_IF_ABSENT == cacheOperation) {
+            request = readKeyValueRequest(cacheOperation, byteBuf);
+        } else {
+            request = new MapCacheRequest(cacheOperation);
+        }
+
+        return request;
+    }
+
+    private MapCacheRequest readKeyRequest(final CacheOperation cacheOperation, final ByteBuf byteBuf) {
+        final Optional<byte[]> key = readBytes(byteBuf);
+        return key.map(bytes -> new MapCacheRequest(cacheOperation, bytes)).orElse(null);
+    }
+
+    private MapCacheRequest readKeyValueRequest(final CacheOperation cacheOperation, final ByteBuf byteBuf) {
+        final MapCacheRequest mapCacheRequest;
+
+        final Optional<byte[]> key = readBytes(byteBuf);
+        if (key.isPresent()) {
```

Comment:
It looks like, for large values (where the first payload read does not fetch the entire payload), there is the potential that the key data would be lost.

Reply:
Thanks for catching that issue @greyp9! I pushed an update with a new test method to exercise this condition, and also adjusted `CacheRequestDecoder` to use `ByteBuf.markReaderIndex()` and `ByteBuf.resetReaderIndex()` to avoid partial reads.